### PR TITLE
Boltzmann interface dev

### PR DESCRIPTION
**Pull request by uvilla**
Developing the interface for the Boltzmann solver interface.

*Key changes*:
- Implemented a `Tps2Boltzmann` class (both c++ and python wrappers) to handle the sharing of information between the TPS and Boltzmann solver. A new section `[boltzmannInterface]` in the input file has been added to pass parameters to the interface.
- Added functionalities in `M2ulPhysS` to push thermodynamic state variables to `Tps2Boltzmann`.
- Modified `QuasiMagnetostaticSolver*` to optionally compute/store the electric field.
- Interpolate the electric field to the fluid mesh in `CycleAvgJouleCoupling`.
- Add a demo of the new interface (including a `BoltzmannMockSolver`) in `tps-time-loos.py`.
- Resolved many memory leaks in `QuasiMagnetostaticSolver*` and `M2ulPhysS`.

*Next milestones*:
- Create a `pytps` python package
- Improve the interface of `Tps2Boltzmann` to include variable name dictionaries and units
- Extend `Tps2Boltzmann` to handle GPU data directly. Dependents on #228 that allow EM solver to run on the build for GPUs
- Implement methods to read back Boltzmann fields in TPS

### CODE REVIEW

**Comment by uvilla** [comment]
@trevilo :

There is a built in that fails in the `VPATH Testing` that I can not reproduce locally.
Other than that this first chuck is ready to be reviewed and merged.

Cheers,

Umberto

**Comment by trevilo** [comment]
> @trevilo :
> 
> There is a built in that fails in the `VPATH Testing` that I can not reproduce locally. Other than that this first chuck is ready to be reviewed and merged.
> 
> Cheers,
> 
> Umberto

Thanks @uvilla.  The failing test is non-deterministic (by design unfortunately) so it fails occasionally.  I'd like to change that, but... that is a battle for another day.  I've triggered it to run again, and I'm guessing it will pass, especially since you couldn't reproduce locally.  If it passes, I'll look this over and get it merged asap.  If not, I'll take a closer look.

**Review by trevilo — CHANGES_REQUESTED** [review]
Hey @uvilla, just two tiny comments/questions.  After that, this looks ready to me.

File: src/M2ulPhyS2Bolzmann.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+// -----------------------------------------------------------------------------------bl-
```
[the remark this answers is not in the document]
Comment (uvilla) [reply src/M2ulPhyS2Bolzmann.cpp]: File has been renamed as suggested.

File: src/cycle_avg_joule_coupling.cpp [hunk 2]
```diff
@@ -352,3 +405,38 @@ void CycleAvgJouleCoupling::solveEnd() {
...
+
+/// Push solver variables to interface
+void CycleAvgJouleCoupling::push(TPS::Tps2Boltzmann &interface) {
+  flow_solver_->push(interface);
+  interface.interpolateFromNativeFES(*efield_, TPS::Tps2Boltzmann::Index::ElectricField);
+}
+
+/// Fetch solver variables from interface
+void CycleAvgJouleCoupling::fetch(TPS::Tps2Boltzmann &interface) {
+  std::cout << "Nothing to be fetched in CycleAvgJouleCoupling" << std::endl;
```
[the remark this answers is not in the document]
Comment (uvilla) [reply src/cycle_avg_joule_coupling.cpp]: Statement has been removed. Thank you for catching that.

File: src/M2ulPhyS2Bolzmann.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (trevilo) [reply src/M2ulPhyS2Bolzmann.cpp]: Doh! I messed up and inadvertently took out the `2`, but what I was trying to point out is that `Bolzmann` should be `Boltzmann`.

**Review by trevilo — APPROVED** [review]

**Pull request merged**